include/lbann/execution_contexts/sgd_execution_context.hpp

```diff
@@ -0,0 +1,119 @@
+////////////////////////////////////////////////////////////////////////////////
+// Copyright (c) 2014-2016, Lawrence Livermore National Security, LLC.
+// Produced at the Lawrence Livermore National Laboratory.
+// Written by the LBANN Research Team (B. Van Essen, et al.) listed in
+// the CONTRIBUTORS file. <lbann-dev@llnl.gov>
+//
+// LLNL-CODE-697807.
+// All rights reserved.
+//
+// This file is part of LBANN: Livermore Big Artificial Neural Network
+// Toolkit. For details, see http://software.llnl.gov/LBANN or
+// https://github.com/LLNL/LBANN.
+//
+// Licensed under the Apache License, Version 2.0 (the "Licensee"); you
+// may not use this file except in compliance with the License.  You may
+// obtain a copy of the License at:
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the license.
+////////////////////////////////////////////////////////////////////////////////
+
+#ifndef LBANN_SGD_EXECUTION_CONTEXT_HPP
+#define LBANN_SGD_EXECUTION_CONTEXT_HPP
+
+#include "lbann/execution_contexts/execution_context.hpp"
+#include <cereal/types/base_class.hpp>
+namespace lbann {
+
+struct sgd_termination_criteria : public termination_criteria {
+  El::Int num_epochs;
+};
+
+
+/** @brief SGD Uses the step to track the Current mini-batch step for
+  *  execution mode.
+  *  @detailed Step counts are not reset after each epoch.
+  */
+class sgd_execution_context final : public execution_context {
+public:
+  /** Constructor. */
+  sgd_execution_context(observing_ptr<trainer> trainer, lbann_comm *comm, execution_mode mode, int mini_batch_size);
```

Comment:
Either `unsigned mini_batch_size` or `size_t mini_batch_size`. (You probably won't have a minibatch larger than 2 billion samples, but you definitely won't have one with -12 samples.

Reply:
I could see a mini-batch with that many samples.

Reply:
2 billion or -12? 😛

Reply:
I meant 2 billion, not 18446744073709551604.

Reply:
What about 4 billion? It's starting to sound like `size_t` is the way to go...

Reply:
For forward compatibility, I would prefer `size_t` unless we know in advance memory usage or performance are an issue that need to be resolved by using a smaller type.

Reply:
In that case:
```suggestion
  sgd_execution_context(observing_ptr<trainer> trainer, lbann_comm *comm, execution_mode mode, size_t mini_batch_size);
```